File: src/requestcompletion/llm/tools/schema_parser.py
"""
JSON schema parsing utilities.

This module contains functions for parsing JSON schemas into Parameter objects
and converting Parameter objects into Pydantic models.
"""

from typing import Dict

from .parameter import Parameter, PydanticParameter


def parse_json_schema_to_parameter(
    name: str, prop_schema: dict, required: bool
) -> Parameter:
    """
    Given a JSON-schema for a property, returns a Parameter or PydanticParameter.
    If prop_schema defines nested properties, this is done recursively.

    Args:
        name: The name of the parameter.
        prop_schema: The JSON schema definition for the property.
        required: Whether the parameter is required.

    Returns:
        A Parameter or PydanticParameter object representing the schema.
    """

    # Handle type as list (union/optional)
    param_type = prop_schema.get("type", None)
    if param_type is None:
        # If no type, try to infer from other keys
        if "properties" in prop_schema:
            param_type = "object"
        elif "items" in prop_schema:
            param_type = "array"
        else:
            param_type = "string"  # fallback

    # Handle special case for number type
    if param_type == "number":
        param_type = "float"

    # Handle type as list (union)
    if isinstance(param_type, list):
        # Convert to python types, e.g. ["string", "null"]
        param_type = [t if t != "null" else "none" for t in param_type]

    description = prop_schema.get("description", "")
    enum = prop_schema.get("enum")
    default = prop_schema.get("default")
    additional_properties = prop_schema.get("additionalProperties", False)

    # Handle references to other schemas
    if "$ref" in prop_schema:
        return PydanticParameter(
            name=name,
            param_type="object",
            description=description,
            required=required,
            properties={},
            additional_properties=additional_properties,
        )

    # Handle allOf (merge schemas)
    if "allOf" in prop_schema:
        # Only handle simple case: allOf with $ref or type
        for item in prop_schema["allOf"]:
            if "$ref" in item:
                # Reference to another schema
                return PydanticParameter(
                    name=name,
                    param_type="object",
                    description=description,
                    required=required,
                    properties={},
                    additional_properties=additional_properties,
                )
            elif "type" in item:
                # Merge type info
                param_type = item["type"]

    # Handle anyOf (union types)
    if "anyOf" in prop_schema:
        # Only handle simple case: anyOf with types
        types_list = []
        for item in prop_schema["anyOf"]:
            t = item.get("type")
            if t:
                types_list.append(t if t != "null" else "none")
        if types_list:
            param_type = types_list

    # Handle nested objects
    if param_type == "object" and "properties" in prop_schema:
        inner_required = prop_schema.get("required", [])
        inner_props = {}
        for inner_name, inner_schema in prop_schema["properties"].items():
            inner_props[inner_name] = parse_json_schema_to_parameter(
                inner_name, inner_schema, inner_name in inner_required
            )
        return PydanticParameter(
            name=name,
            param_type="object",
            description=description,
            required=required,
            properties=inner_props,
            additional_properties=additional_properties,
        )

    # Handle arrays, potentially with nested objects
    elif param_type == "array" and "items" in prop_schema:
        items_schema = prop_schema["items"]
        if items_schema.get("type") == "object" and "properties" in items_schema:
            inner_required = items_schema.get("required", [])
            inner_props = {}
            for inner_name, inner_schema in items_schema["properties"].items():
                inner_props[inner_name] = parse_json_schema_to_parameter(
                    inner_name, inner_schema, inner_name in inner_required
                )
            return PydanticParameter(
                name=name,
                param_type="array",
                description=description,
                required=required,
                properties=inner_props,
                additional_properties=additional_properties,
            )
        else:
            return Parameter(
                name=name,
                param_type="array",
                description=description,
                required=required,
                enum=enum,
                default=default,
                additional_properties=additional_properties,
            )
    else:
        return Parameter(
            name=name,
            param_type=param_type,
            description=description,
            required=required,
            enum=enum,
            default=default,
            additional_properties=additional_properties,
        )


def parse_model_properties(schema: dict) -> Dict[str, Parameter]:  # noqa: C901
    """
    Given a JSON schema (usually from BaseModel.model_json_schema()),
    returns a dictionary mapping property names to Parameter objects.

    Args:
        schema: The JSON schema to parse.

    Returns:
        A dictionary mapping property names to Parameter objects.
    """
    result = {}
    required_fields = schema.get("required", [])

    # First, process any $defs (nested model definitions)
    defs = schema.get("$defs", {})
    nested_models = {}

    for def_name, def_schema in defs.items():
        # Parse each nested model definition
        nested_props = {}
        nested_required = def_schema.get("required", [])

        for prop_name, prop_schema in def_schema.get("properties", {}).items():
            nested_props[prop_name] = parse_json_schema_to_parameter(
                prop_name, prop_schema, prop_name in nested_required
            )

        nested_models[def_name] = {
            "properties": nested_props,
            "required": nested_required,
        }

    # Process main properties
    for prop_name, prop_schema in schema.get("properties", {}).items():
        # Check if this property references a nested model
        if "$ref" in prop_schema:
            ref = prop_schema["$ref"]
            if ref.startswith("#/$defs/"):
                model_name = ref[len("#/$defs/") :]
                if model_name in nested_models:
                    # Create a PydanticParameter with the nested model's properties
                    result[prop_name] = PydanticParameter(
                        name=prop_name,
                        param_type="object",
                        description=prop_schema.get("description", ""),
                        required=prop_name in required_fields,
                        properties=nested_models[model_name]["properties"],
                    )
                    continue
        elif "allOf" in prop_schema:
            for item in prop_schema.get("allOf", []):
                if "$ref" in item:
                    # Extract the model name from the reference
                    ref = item["$ref"]
                    if ref.startswith("#/$defs/"):
                        model_name = ref[len("#/$defs/") :]
                        if model_name in nested_models:
                            # Create a PydanticParameter with the nested model's properties
                            result[prop_name] = PydanticParameter(
                                name=prop_name,
                                param_type="object",
                                description=prop_schema.get("description", ""),
                                required=prop_name in required_fields,
                                properties=nested_models[model_name]["properties"],
                            )
                            break

        # If not already processed as a reference
        if prop_name not in result:
            # Get the correct type from the schema
            param_type = prop_schema.get("type", "object")

            # Handle special case for number type
            if "type" in prop_schema and prop_schema["type"] == "number":
                param_type = "float"

            # Create parameter with the correct type
            if param_type == "object" and "properties" in prop_schema:
                inner_required = prop_schema.get("required", [])
                inner_props = {}
                for inner_name, inner_schema in prop_schema["properties"].items():
                    inner_props[inner_name] = parse_json_schema_to_parameter(
                        inner_name, inner_schema, inner_name in inner_required
                    )
                result[prop_name] = PydanticParameter(
                    name=prop_name,
                    param_type=param_type,
                    description=prop_schema.get("description", ""),
                    required=prop_name in required_fields,
                    properties=inner_props,
                )
            else:
                result[prop_name] = parse_json_schema_to_parameter(
                    prop_name, prop_schema, prop_name in required_fields
                )

    return result

File: dotnet/src/Railtracks.Core/Llm/Tools/SchemaParser.cs
namespace Railtracks.Llm.Tools;

using JsonObject = IReadOnlyDictionary<string, object?>;

/// <summary>
/// Parses JSON-schema property dictionaries into <see cref="Parameter"/> instances.
/// Ported from the Python <c>schema_parser</c> module. Schema objects are represented as
/// <c>IReadOnlyDictionary&lt;string, object?&gt;</c> (matching a deserialized JSON object).
/// </summary>
public static class SchemaParser
{
    /// <summary>Parse a JSON-schema property dict into the appropriate <see cref="Parameter"/> subtype.</summary>
    public static Parameter ParseJsonSchemaToParameter(string name, JsonObject schema, bool required)
    {
        var paramType = ExtractParamType(schema);
        var description = GetString(schema, "description") ?? string.Empty;
        var @enum = GetList(schema, "enum");
        var defaultPresent = schema.ContainsKey("default");
        var @default = schema.TryGetValue("default", out var d) ? d : null;
        var additionalProperties = GetBool(schema, "additionalProperties");

        if (schema.ContainsKey("$ref"))
            return new RefParameter(name, (string)schema["$ref"]!, description, required);

        if (schema.ContainsKey("allOf"))
        {
            var (result, updatedType) = HandleAllOf(name, schema, required, description, additionalProperties);
            if (result is not null)
                return result;
            if (updatedType is not null)
                paramType = updatedType;
        }

        if (schema.ContainsKey("anyOf"))
            return HandleAnyOf(name, schema, required, description, @default, defaultPresent);

        if ((Equals(paramType, "object") || (paramType is List<string> l && l.Contains("object")))
            && schema.ContainsKey("properties"))
        {
            return HandleObject(name, schema, required, description, additionalProperties);
        }

        if (Equals(paramType, "array") && schema.ContainsKey("items"))
            return HandleArray(name, schema, required, description, @default, additionalProperties);

        if (paramType is List<string> types)
        {
            var options = types.Select(t => MapTokenToParameter(t)).ToList();
            return new UnionParameter(name, options, description, required, @default, @enum, defaultPresent);
        }

        return new Parameter(name, description, required, @default, @enum, defaultPresent, paramType);
    }

    /// <summary>Returns the top-level properties of a (pydantic-style) JSON schema as parameters.</summary>
    public static List<Parameter> ParseModelProperties(JsonObject schema)
    {
        var requiredFields = GetStringList(schema, "required");
        var nestedModels = ParseModelDefs(GetObject(schema, "$defs"));
        return ParseMainProperties(GetObject(schema, "properties"), requiredFields, nestedModels);
    }

    // ---- helpers ported from the Python module ----

    private static object ExtractParamType(JsonObject schema)
    {
        object? paramType = schema.TryGetValue("type", out var t) ? t : null;

        if (paramType is null)
        {
            if (schema.ContainsKey("properties")) return "object";
            if (schema.ContainsKey("items")) return "array";
            return "object";
        }

        if (paramType is IEnumerable<object?> list)
        {
            return list.Select(x => x as string == "null" ? "none" : (string)x!).ToList();
        }

        return paramType;
    }

    private static (Parameter?, object?) HandleAllOf(
        string name, JsonObject schema, bool required, string description, bool additionalProperties)
    {
        object? paramType = null;
        foreach (var itemObj in GetObjectList(schema, "allOf"))
        {
            if (itemObj.ContainsKey("$ref"))
            {
                return (new ObjectParameter(name, new List<Parameter>(), description, required, additionalProperties), null);
            }

            if (itemObj.TryGetValue("type", out var ty))
                paramType = ty;
        }

        return (null, paramType);
    }

    private static Parameter HandleAnyOf(
        string name, JsonObject schema, bool required, string description, object? @default, bool defaultPresent)
    {
        var options = new List<Parameter>();
        var idx = 0;
        foreach (var optionSchema in GetObjectList(schema, "anyOf"))
            options.Add(ParseJsonSchemaToParameter($"{name}_option_{idx++}", optionSchema, true));

        var flattened = new List<Parameter>();
        foreach (var opt in options)
        {
            if (opt is UnionParameter union)
                flattened.AddRange(union.Options);
            else
                flattened.Add(opt);
        }

        if (flattened.Count == 1)
            return flattened[0];

        return new UnionParameter(name, flattened, description, required, @default, null, defaultPresent);
    }

    private static ObjectParameter HandleObject(
        string name, JsonObject schema, bool required, string description, bool additionalProperties)
    {
        var innerRequired = GetStringList(schema, "required");
        var innerProps = GetObject(schema, "properties")
            .Select(kv => ParseJsonSchemaToParameter(kv.Key, AsObject(kv.Value), innerRequired.Contains(kv.Key)))
            .ToList();

        return new ObjectParameter(
            name,
            innerProps,
            string.IsNullOrEmpty(description) ? GetString(schema, "description") : description,
            required,
            GetBool(schema, "additionalProperties", additionalProperties),
            schema.TryGetValue("default", out var def) ? def : null);
    }

    private static ArrayParameter HandleArray(
        string name, JsonObject schema, bool required, string description, object? @default, bool additionalProperties)
    {
        var itemsSchema = AsObject(schema["items"]);
        var maxItems = schema.TryGetValue("maxItems", out var mi) && mi is not null ? Convert.ToInt32(mi) : (int?)null;

        Parameter itemParam;
        if (Equals(GetString(itemsSchema, "type"), "object") && itemsSchema.ContainsKey("properties"))
        {
            var innerRequired = GetStringList(itemsSchema, "required");
            itemParam = new ObjectParameter(
                $"{name}_item",
                GetObject(itemsSchema, "properties")
                    .Select(kv => ParseJsonSchemaToParameter(kv.Key, AsObject(kv.Value), innerRequired.Contains(kv.Key)))
                    .ToList(),
                GetString(itemsSchema, "description") ?? string.Empty,
                true,
                GetBool(itemsSchema, "additionalProperties"));
        }
        else
        {
            itemParam = new Parameter(
                $"{name}_item",
                GetString(itemsSchema, "description") ?? string.Empty,
                true,
                itemsSchema.TryGetValue("default", out var idef) ? idef : null,
                GetList(itemsSchema, "enum"),
                paramType: GetString(itemsSchema, "type") ?? "string");
        }

        return new ArrayParameter(name, itemParam, description, required, @default, maxItems, additionalProperties);
    }

    private static Dictionary<string, List<Parameter>> ParseModelDefs(JsonObject defs)
    {
        var result = new Dictionary<string, List<Parameter>>();
        foreach (var (defName, defSchemaObj) in defs)
        {
            var defSchema = AsObject(defSchemaObj);
            var nestedRequired = GetStringList(defSchema, "required");
            var nestedProps = GetObject(defSchema, "properties")
                .Select(kv => ParseJsonSchemaToParameter(kv.Key, AsObject(kv.Value), nestedRequired.Contains(kv.Key)))
                .ToList();
            result[defName] = nestedProps;
        }

        return result;
    }

    private static List<Parameter> ParseMainProperties(
        JsonObject properties, IReadOnlyCollection<string> requiredFields, IReadOnlyDictionary<string, List<Parameter>> nestedModels)
    {
        var result = new List<Parameter>();
        foreach (var (propName, propSchemaObj) in properties)
        {
            var propSchema = AsObject(propSchemaObj);

            if (propSchema.ContainsKey("$ref"))
            {
                var refObj = HandleRefProperty(propName, propSchema, requiredFields, nestedModels);
                if (refObj is not null) { result.Add(refObj); continue; }
            }

            if (propSchema.ContainsKey("allOf"))
            {
                var allOfObj = HandleAllOfProperty(propName, propSchema, requiredFields, nestedModels);
                if (allOfObj is not null) { result.Add(allOfObj); continue; }
            }

            var paramType = GetString(propSchema, "type") ?? "object";
            if (paramType == "object" && propSchema.ContainsKey("properties"))
                result.Add(HandleObjectProperty(propName, propSchema, requiredFields));
            else
                result.Add(ParseJsonSchemaToParameter(propName, propSchema, requiredFields.Contains(propName)));
        }

        return result;
    }

    private static ObjectParameter? HandleRefProperty(
        string propName, JsonObject propSchema, IReadOnlyCollection<string> requiredFields,
        IReadOnlyDictionary<string, List<Parameter>> nestedModels)
    {
        var refPath = (string)propSchema["$ref"]!;
        if (refPath.StartsWith("#/$defs/", StringComparison.Ordinal))
        {
            var modelName = refPath["#/$defs/".Length..];
            if (nestedModels.TryGetValue(modelName, out var props))
            {
                return new ObjectParameter(propName, props, GetString(propSchema, "description") ?? string.Empty,
                    requiredFields.Contains(propName));
            }
        }

        return null;
    }

    private static ObjectParameter? HandleAllOfProperty(
        string propName, JsonObject propSchema, IReadOnlyCollection<string> requiredFields,
        IReadOnlyDictionary<string, List<Parameter>> nestedModels)
    {
        foreach (var item in GetObjectList(propSchema, "allOf"))
        {
            if (item.ContainsKey("$ref"))
            {
                var refPath = (string)item["$ref"]!;
                if (refPath.StartsWith("#/$defs/", StringComparison.Ordinal))
                {
                    var modelName = refPath["#/$defs/".Length..];
                    if (nestedModels.TryGetValue(modelName, out var props))
                    {
                        return new ObjectParameter(propName, props, GetString(propSchema, "description") ?? string.Empty,
                            requiredFields.Contains(propName));
                    }
                }
            }
        }

        return null;
    }

    private static ObjectParameter HandleObjectProperty(
        string propName, JsonObject propSchema, IReadOnlyCollection<string> requiredFields)
    {
        var innerRequired = GetStringList(propSchema, "required");
        var innerProps = GetObject(propSchema, "properties")
            .Select(kv => ParseJsonSchemaToParameter(kv.Key, AsObject(kv.Value), innerRequired.Contains(kv.Key)))
            .ToList();

        return new ObjectParameter(propName, innerProps, GetString(propSchema, "description") ?? string.Empty,
            requiredFields.Contains(propName), GetBool(propSchema, "additionalProperties"));
    }

    // Replicates the Python `param_from_python_type` behavior for a schema-type token (only
    // "none" is a recognized key; anything else maps to object — matching from_python_type).
    private static Parameter MapTokenToParameter(string token)
    {
        var type = token == "none" ? ParameterType.None : ParameterType.Object;
        return new Parameter(string.Empty, paramType: type.Value());
    }

    // ---- dictionary access helpers ----

    private static JsonObject AsObject(object? value) =>
        value as JsonObject
        ?? (value is IDictionary<string, object?> d ? new Dictionary<string, object?>(d) : null)
        ?? new Dictionary<string, object?>();

    private static string? GetString(JsonObject schema, string key) =>
        schema.TryGetValue(key, out var v) ? v as string : null;

    private static bool GetBool(JsonObject schema, string key, bool fallback = false) =>
        schema.TryGetValue(key, out var v) && v is bool b ? b : fallback;

    private static IReadOnlyList<object?>? GetList(JsonObject schema, string key) =>
        schema.TryGetValue(key, out var v) && v is IEnumerable<object?> e ? e.ToList() : null;

    private static List<string> GetStringList(JsonObject schema, string key) =>
        schema.TryGetValue(key, out var v) && v is IEnumerable<object?> e
            ? e.Select(x => x?.ToString() ?? string.Empty).ToList()
            : new List<string>();

    private static JsonObject GetObject(JsonObject schema, string key) =>
        schema.TryGetValue(key, out var v) ? AsObject(v) : new Dictionary<string, object?>();

    private static IEnumerable<JsonObject> GetObjectList(JsonObject schema, string key) =>
        schema.TryGetValue(key, out var v) && v is IEnumerable<object?> e
            ? e.Select(AsObject)
            : Enumerable.Empty<JsonObject>();
}
